Symbols perf: use sqlite3

chrismwendt opened the pull request:
## Background

Previously, the symbols service was pretty slow on big repositories (linear in the repository size). See https://github.com/sourcegraph/sourcegraph/issues/2257 for analysis.

**GIF of kubernetes/kubernetes code intel before this PR**

![2019-02-18 01 25 15](https://user-images.githubusercontent.com/1387653/52944531-71a52b80-3324-11e9-8c81-80862670ac89.gif)

This vastly improves the speed of searches (it uses sqlite3 and indexes the symbol name column for nearly instant queries). This has no effect on the speed of initial indexing of symbols.

**GIF of kubernetes/kubernetes code intel after this PR**

![2019-02-18 01 19 54](https://user-images.githubusercontent.com/1387653/52944550-81bd0b00-3324-11e9-87ca-11a6e1493411.gif)

**Benchmark results**

```
$ env CTAGS_COMMAND=$PWD/cmd/symbols/universal-ctags-dev go test -v -bench=. -test.run=sdfsfd ./cmd/symbols/internal/symbols
goos: darwin
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols
BenchmarkSearch/indexing_go-langserver@391-8         	      10	 140234483 ns/op
BenchmarkSearch/indexing_moby@6e5-8                  	       1	7930319865 ns/op
BenchmarkSearch/searching_go-langserver@391_^sortedImportRecord$-8         	   30000	     44872 ns/op
BenchmarkSearch/searching_go-langserver@391_1234doesnotexist1234-8         	    5000	    281762 ns/op
BenchmarkSearch/searching_moby@6e5_^fsCache$-8                             	   20000	     80045 ns/op
BenchmarkSearch/searching_moby@6e5_1234doesnotexist1234-8                  	      30	  46515624 ns/op
PASS
ok  	github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols	36.963s
```

I forced the non exact regexes to be run over all symbols by using `1234doesnotexist1234` as the regex. Exact matches (`^...$`) are much faster and scale better than non exact matches that don't exist.

The current implementation takes ~40ms to search for `1234doesnotexist1234` in moby/moby, and takes ~46ms using sqlite3, which I think is acceptable.

**Memory usage improvements**

This also reduces the memory consumption (it streams symbols from the ctags process to sqlite3 without needing to build up a giant `[]Symbol` in Go). This should reduce the frequency of the symbols service getting OOM killed (cc @ggilmore, just FYI).

- Memory usage while indexing kubernetes/kubernetes **before this PR**: ~200 MB
- Memory usage while indexing kubernetes/kubernetes **after this PR**: ~26 MB

**TypeScript build script**

This also uses TypeScript for a build script. It's much safer than shell, and I'm curious if anyone has thoughts on this.

## Review

This is one of the largest changes I've made, and I'd be happy to do a video chat 🎥 to make the review process go quickly and smoothly if you want.

Here are the areas I'd like review on (feel free to review anything, these are just my guesses for what you will want to look at):

- @keegancsmith 
  - Correct usage of concurrency primitives, especially in the functions that return channels for streaming symbols (I'm pretty new to this area of Go, so there is a high probability I've done things that are plain wrong)
  - Correct usage of sqlite3, sqlf, and sqlx
  - Database schema
- @slimsag 
  - Dev build system changes (mostly `build.ts`, some changes to Procfile and other shell scripts)
  - General Go idioms, control flow, organization of functions, methods, etc.
- @felixfbecker 
  - Usage of ts-node, `./dev/ts-script`, and `cmd/symbols/build.ts`
- Anyone: am I missing anything that you'd expect to see?

I'm hoping to address feedback until both @slimsag and @keegancsmith are confident that this is a robust implementation and are happy with the result before merging.

## TODO

- ~~Add tests~~ they already exist
- [x] Add a few more comments, minor refactoring
- [x] Have a few people run this branch locally to see what parts of the build break (missing libs, missing sqlite3, etc.), since my machine is dirty (both macOS and Linux)

## Issue references

Resolves https://github.com/sourcegraph/sourcegraph/issues/2257
Fixes https://github.com/sourcegraph/sourcegraph/issues/1625

keegancsmith commented:
You need to somehow make go test work

```
$ go test -v ./cmd/symbols/... ./pkg/symbols/...
go: downloading github.com/jmoiron/sqlx v1.2.0
go: downloading golang.org/x/text v0.3.1-0.20180807135948-17ff2d5776d2
?       github.com/sourcegraph/sourcegraph/cmd/symbols  [no test files]
=== RUN   TestParser
--- SKIP: TestParser (0.00s)
    parser_test.go:13: command not in PATH: universal-ctags
PASS
ok      github.com/sourcegraph/sourcegraph/cmd/symbols/internal/pkg/ctags       0.007s
=== RUN   TestService
--- FAIL: TestService (0.01s)
panic: Can't find the libsqlite3-pcre library because LIBSQLITE3_PCRE was not set and libsqlite3-pcre.* doesn't exist at the root of the repository. Try building it with `./dev/ts-script cmd/symbols/build.ts buildLibsqlite3Pcre`. [recovered]
        panic: Can't find the libsqlite3-pcre library because LIBSQLITE3_PCRE was not set and libsqlite3-pcre.* doesn't exist at the root of the repository. Try building it with `./dev/ts-script cmd/symbols/build.ts buildLibsqlite3Pcre`.

goroutine 39 [running]:
testing.tRunner.func1(0xc000318100)
        /usr/local/Cellar/go/1.11.5/libexec/src/testing/testing.go:792 +0x387
panic(0x4809700, 0x49bda50)
        /usr/local/Cellar/go/1.11.5/libexec/src/runtime/panic.go:513 +0x1b9
github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols.TestService(0xc000318100)
        /Users/keegan/go/src/github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols/service_test.go:40 +0xbf1
testing.tRunner(0xc000318100, 0x4949450)
        /usr/local/Cellar/go/1.11.5/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
        /usr/local/Cellar/go/1.11.5/libexec/src/testing/testing.go:878 +0x35c
FAIL    github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols 0.032s
?       github.com/sourcegraph/sourcegraph/pkg/symbols  [no test files]
?       github.com/sourcegraph/sourcegraph/pkg/symbols/protocol [no test files]
```

Requiring us to set an environment variable will break a bunch of workflows. It should either just work, or skip if it can't work (except fail on CI. You can check if the envvar CI is non-empty for that).

keegancsmith commented on cmd/symbols/build.ts:
```diff
@@ -0,0 +1,233 @@
+    run('docker', 'build', '--quiet', '-f', 'cmd/symbols/Dockerfile', '-t', dockerImageName, dockerBuildContext)
+    shelljs.rm('-rf', dockerBuildContext)
+    console.log(`Building the ${dockerImageName} Docker image... done`)
+}
```
I'm not really understanding why this is better than a shell script? Personally this seems harder to read than a shell script, since it doesn't contain any complicated logic. We also lose things like being able to do `bash -x` to understand exactly what commands are running.

chrismwendt replied:
The lack of `set -x` is a fair point, but I much prefer TypeScript for the type safety, autocomplete, objects, npm package ecosystem, stack traces, and nice code intel. I'm also not nearly as well versed in bash as you are 😛

felixfbecker replied:
I get that but instead of type safety and autocomplete this script uses shellJS so it's really just a glorified shell script?

slimsag replied:
I am not convinced this is resolved.

chrismwendt replied:
What do you guys say we try it out and see if any issues arise? I'll be on the hook for problems, fixes, and related changes in the future.

keegancsmith commented on cmd/symbols/internal/symbols/parse.go:
```diff
@@ -112,18 +113,21 @@ func (s *Service) parseUncached(ctx context.Context, repo api.RepoName, commitID
+					err = callback(entryToSymbol(e))
+					if err != nil {
+						cancel()
+						log15.Error(err.Error())
```
```suggestion
						log15.Error("Failed to add symbol", "symbol", e, "error", err)
```

keegancsmith commented on cmd/symbols/internal/symbols/parse.go:
```diff
@@ -112,18 +113,21 @@ func (s *Service) parseUncached(ctx context.Context, repo api.RepoName, commitID
+					totalSymbols++
+					err = callback(entryToSymbol(e))
+					if err != nil {
+						cancel()
```
you don't need to call cancel, it will be called by the above defer when you return.

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -73,34 +77,52 @@ func (s *Service) search(ctx context.Context, args protocol.SearchArgs) (result
-	const maxFirst = 500
-	if args.First < 0 || args.First > maxFirst {
-		args.First = maxFirst
+	db, err := sqlx.Open("sqlite3_with_pcre", dbFile)
```
it will be interesting to test how this performs on sourcegraph or dogfood when you search across lots of repositories at once.

chrismwendt commented:
I changed the test so that it only panics in CI.

CI builds libsqlite3-pcre prior to running the tests in gen-pipeline.go.

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -73,34 +77,52 @@ func (s *Service) search(ctx context.Context, args protocol.SearchArgs) (result
+
+func (s *Service) getDBFile(ctx context.Context, args protocol.SearchArgs) (string, error) {
+	diskcacheFile, err := s.cache.Open(ctx, fmt.Sprintf("%d-%s@%s", symbolsDbVersion, args.Repo, args.CommitID), func(fetcherCtx context.Context) (io.ReadCloser, error) {
+		tempDBFile, err := ioutil.TempFile("", "")
```
TEMP_DIR doesn't have a lot of space in production. So this will lead to issues. diskcache doesn't expect you to be using an intermediate disk location (it abstracts that away). Ideally you want to write to the same temp location (on the SSD), and let diskcache do the atomic renaming/etc.

I would suggest adding a new method which instead passes in the temporary file location for you to write too.

chrismwendt replied:
Good idea, I added a new method `OpenWithFile` that provides the temp file path to the fetcher. This avoids writing to TEMP_DIR, and also avoids the extra file copy.

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +131,210 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+				rhs = strings.ToLower(rhs)
+				column = column + "lowercase"
+			}
+			conditions = append(conditions, sqlf.Sprintf(column+" = %s", rhs))
```
I'm surprised sqlite can't detect this. Have you tested if this makes a performance difference?

chrismwendt replied:
Yes, I tested this and sqlite did not optimize regexes of the form `^...$`.

keegancsmith replied:
Yeah that makes sense after I looked at the code for SQLite-PCRE. :+1:

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +131,210 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+	conds = append(conds, negateAll(mkConds("path", args.ExcludePattern))...)
+
+	var sqlQuery *sqlf.Query
+	if len(conds) == 0 {
```
a common pattern is to always have the condition ("TRUE") in the list so you don't need to special case no conditions.

chrismwendt replied:
Unfortunately, adding ("TRUE") breaks the semantics when using negation to implement the excluded patterns (it ends up querying for `NOT TRUE`).

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +131,210 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+				"INSERT INTO symbols %s VALUES %s",
+				"( name,  namelowercase,  path,  pathlowercase,  line,  kind,  language,  parent,  parentkind,  signature,  pattern,  filelimited)",
+				"(:name, :namelowercase, :path, :pathlowercase, :line, :kind, :language, :parent, :parentkind, :signature, :pattern, :filelimited)"),
+			&symbolInDBValue)
```
you should experiment with more batched ways of inserting the entries. Each of these is a cgo call, which is quite expensive. Additionally I'd expect a prepared query would also be fast.

chrismwendt replied:
Writing via prepared statement only speeds this up by ~2% when there's 1 ctags process and ~10% with 8 ctags processes (on my mac with 8 CPUs).

Do you know of other ways to batch insert entries? I read about using a transaction, which was definitely a huge speed boost (something like 10x faster).

chrismwendt commented:
Thanks for the review! 🙇 

Is it at all likely that checked-in libs would not work on different versions of macOS or Linux? That's my main concern with checking in the libs. The build script only compiles libsqlite3-pcre if it's not present at the repository root (so that's only once for each developer machine).

A musl cross compiler is only necessary to build the `symbols` executable on the host prior to putting it into the Docker build context for the dev-symbols image that has the `universal-ctags` executable that is used by `universal-ctags-dev`, so I could drop that requirement in dev by factoring out universal-ctags as its own Docker image and only building that in dev. So that would mean 2 images: `dev-symbols` and `universal-ctags`. Is that what you would do to eliminate the requirement?

Is there an equivalent to `LD_LIBRARY_PATH` on macOS? If I put libsqlite3-pcre in `/usr/local/lib`, I'd need to implement logic to loop over the components of `LD_LIBRARY_PATH` and look for the file (which might not be named exactly libsqlite3-pcre). It seems easier and more reliable to pass `LIBSQLITE3_PCRE`.

felixfbecker commented:
I am always in favour of not checking in dependencies.

Is it possible to get that lib with a package manager?

keegancsmith commented:
This is a C dep, so the go package manager won't fetch it. Using CGO is a choose your own adventure unfortunately.

I would suggest we move the build logic into the dockerfile, and we make it self contained. That way we don't need cross compiling.

In the case of running tests we skip if the dynamic lib is not found. Or we could try use the same approach as go-sqlite does and vendor the sqlite-pcre code in the repo (this is common way to distribute sqlite).

We also need to ensure the `server` image works with symbols + sqlite.

Sorry I didn't realise Mac didn't use LD_LIBRARY_PATH, I've never done C/C++ dev on a mac, only linux.

Making our `go test` or build environment more complicated is a blocker for merging this, unless there is no other alternative.

keegancsmith commented:
I'd also be interested in using something like https://godoc.org/crawshaw.io/sqlite instead, since its motivation for existing are pretty great. See this blog post for motivation https://crawshaw.io/blog/go-and-sqlite

felixfbecker commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +134,212 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+		if strings.HasPrefix(regex, "^") && strings.HasSuffix(regex, "$") {
+			// It looks like the user is asking for exact matches, so use `=`
+			// for speed. Checking for `^...$` isn't 100% accurate, but it
+			// covers 99.9% of cases.
```
I don't think this is good. I can imagine a lot of use cases where the user provides a regex search query that start with `^` and end with `$`. That doesn't mean it is an exact string match search. E.g. consider `^register.*Provider$` to find all provider registration methods in our extension host, but NOT return e.g. a method called `unregisterHoverProvider` or a Map `registerProviderFunctions`. The ^ and $ are needed to enforce that.

You should regex-escape the content between `^` and `$` with `regex.QuoteMeta()` and check if the escaped version is equal to the raw version. If yes, you can do a fast equality search, if not, use regex.

chrismwendt replied:
Good point, this was a shortcut. Fixed.

chrismwendt commented:
> I would suggest we move the build logic into the dockerfile, and we make it self contained. That way we don't need cross compiling.

I actually wanted to go this route, too, until I ran into a few issues:

- Copying the build context of sourcegraph/sourcegraph takes a long time (something like a minute). I suppose we could `cp $(git ls-files) /tmp/context` and build from there, but then you don't get working tree changes, and copying those over is harder.
- We depend on a ton of Go modules that need to be compiled each time the docker build runs, unless they were copied into the build context too (slow, and that wouldn't work because they're all platform-specific, right?)
- Making changes to any files invalidates the Docker layer, and causes Docker to rebuild the world.

It seemed like way too much of a headache than it was worth, considering all we need is a musl cross compiler.

> vendor the sqlite-pcre code in the repo (this is common way to distribute sqlite).

We'd still need to compile it, right? That only avoids downloading the source code, which only happens once per developer machine.

felixfbecker commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +134,212 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+			// It looks like the user is asking for exact matches, so use `=`
+			// for speed. Checking for `^...$` isn't 100% accurate, but it
+			// covers 99.9% of cases.
+			symbolName := strings.TrimSuffix(strings.TrimPrefix(regex, "^"), "$")
```
Slice would probably be faster because it doesn't need to check again if the string starts/ends with the characters and doesn't need a function call
```suggestion
			symbolName := regex[1:len(regex)-1]
```

chrismwendt replied:
For each query (e.g. one basic code intel hover), this is called once for the query, once for the file exclude pattern, and once for each file include pattern (and usually there is 1 of those), which makes me think this is not a performance concern. The intent to strip the `^...$` is clearer when written using `Trim*` than with `regex[1:len(regex)-1]`.

chrismwendt commented on cmd/symbols/build.ts:
```diff
@@ -0,0 +1,233 @@
+function runArrayOrShell(command: string[] | string): shelljs.ExecOutputReturnValue {
+    const oldSilent = shell.config.silent
+    shell.config.silent = true
+    const result = typeof command === 'string' ? shell.exec(command) : shell[command[0]](...command.slice(1))
```
It looks like it's easier to get the exit code and stdout using shelljs.

felixfbecker replied:
execa returns the output as a string and will throw on a non-zero exit code. How is shell JS easier?

chrismwendt replied:
Oh yup, you're right.

chrismwendt commented on cmd/symbols/build.ts:
```diff
@@ -0,0 +1,233 @@
+    run('git', 'checkout', 'c98da412b431edb4db22d3245c99e6c198d49f7a')
+    runShell(buildCommand)
+    shelljs.popd('-q')
+    shelljs.rm('-rf', sqlite3PcreRepositoryDirectory)
```
The only alternative I can think of is rimraf, and I'd bet that `shelljs.rm('-rf', ...)` is more familiar to teammates.

It's easy to find one part of the script that's not type checked and say that part is no better than shell. If I wrote a shell script instead, I'd lose nice functions, objects, and types elsewhere in this script.

slimsag commented on cmd/symbols/build.ts:
```diff
@@ -0,0 +1,270 @@
+// tslint:disable:no-string-literal
```
I am not in favor of doing build scripts in TS/JS like this, honestly. I am worried that this is going to cause our build scripts to quickly become very complex (and honestly, looking at this file, I am not sure how I would interact with it efficiently).

I think a better approach would be sticking with bash and, if you *really* have a need for it, calling out to a script for some exceptional circumstances where using bash is overly tedious (such as writing a heavy amount of functions, for loops, etc).

slimsag commented:
Apologies for not looking into this sooner, which I suspect would've spared you some time. I mentioned earlier on Slack that I would help out with getting this working nicely in dev / CI.

If we can switch to https://github.com/crawshaw/sqlite in fact, we won't need to install the sqlite dependency at all anywhere because it bundles in the source. That is an awesome approach which I am a huge fan of. This means we have a "you just need Go + a C compiler and you're good to go" build scenario, instead of needing everyone to install libsqlite on their own everywhere.

chrismwendt commented:
@slimsag We need the sqlite3 package anyway to [build libsqlite3-pcre](https://github.com/ralight/sqlite3-pcre/blob/c98da412b431edb4db22d3245c99e6c198d49f7a/Makefile#L4) for regex support. https://github.com/crawshaw/sqlite provide any other benefits that make it worth switching now, or could that be done later?

slimsag commented:
I think it would still be very valuable.

For now, you could make it such that dev mode disables pcre support (this would mean you could still start a sourcegraph dev server and do everything, including basic code intel, except symbols regex search).

Also, since libsqlite3-pcre is literally just a plugin, it would not be very hard to precompile and upload it to e.g. a gcloud bucket and download it on dev servers / CI automatically in the future.

I agree it isn't strictly needed to merge this change, but you are definitely in the best position to switch this codebase over to https://github.com/crawshaw/sqlite and that I think will have clear and substantial benefits for us. It will be harder for someone else to do it in the future.

chrismwendt commented:
Considering that running a query on the `name` index takes ~0.01s even on big monorepos, it doesn't seem like our use case would benefit much from crawshaw/sqlite. It'll be a while before we run into concurrency problems. The indexing side is the slow part now, and that's single threaded, so it wouldn't benefit from anything from crawshaw/sqlite as far as I can tell.

Would we be OK with dev diverging from CI like that if we disabled pcre in dev? I'd rather not set that precedent. The build script now tells you how to install the dependencies for macOS/Linux, or links to https://github.com/sourcegraph/sourcegraph/blob/master/doc/dev/local_development.md#step-2-install-dependencies if the OS is unknown:

```
$ ./dev/ts-script cmd/symbols/build.ts execute
Missing sqlite dependencies.
Install them by running `brew install pkg-config sqlite pcre FiloSottile/musl-cross/musl-cross`
```

@keegancsmith I fixed the benchmark and added a query for `^container$` on moby/moby. There's not much for benchmarks there, but it can serve as a base.

keegancsmith commented:
> Considering that running a query one the `name` index takes ~0.01s even on big monorepos, it doesn't seem like our use case would benefit much from crawshaw/sqlite

Is that for a regex as well? Additionally the main performance concerns are two fold:
- Creating the database as quickly as possible
- Concurrent searching across many databases (`type:symbol` search queries)

My guess is crawshaw sqlite will be much better at this, and shouldn't be much of a change to the current code. Given this change is for performance and (I think at least) it is a relatively low friction change I'd be very interested in how it performs. However, you have a good point that this perf is really good (and we could always explore crawshaw sqlite if we need to). I think checking out his repo may help us reduce the need for fetching the pcre extension. His repo is an example of bundling in extensions:
> The optional SQLite3 compiled in are: FTS5, RTree, JSON1, Session

I'll take another look over the code and try it out again. Expect an approval soon :P

chrismwendt commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -9,8 +9,15 @@ import (
+
+	"github.com/sourcegraph/sourcegraph/pkg/api"
+
+	"github.com/jmoiron/sqlx"
```
Mostly for [named queries](http://jmoiron.github.io/sqlx/#namedParams) so that I can write:

```
	insertStatement, err := transaction.PrepareNamed(
		fmt.Sprintf(
			"INSERT INTO symbols %s VALUES %s",
			"( name,  namelowercase,  path,  pathlowercase,  line,  kind,  language,  parent,  parentkind,  signature,  pattern,  filelimited)",
			"(:name, :namelowercase, :path, :pathlowercase, :line, :kind, :language, :parent, :parentkind, :signature, :pattern, :filelimited)"))
	_, err := insertStatement.Exec(&symbolInDBValue)
```

Instead of positional arguments that are easy to mix up `(?, ?, ?, ?, ...)`

chrismwendt commented on pkg/diskcache/cache.go:
```diff
@@ -49,9 +49,33 @@ type File struct {
+		if err != nil {
+			return err
+		}
+		file, err := os.OpenFile(path, os.O_WRONLY, 0600)
```
`OpenWithPath` creates the file and passes the file path to `FetcherWithPath`, so `FetcherWithPath` needs to open the file in order to write to it.

I decided against passing an `os.File` because:

- `sqlx.Open("sqlite3_with_pcre", dbFile)` requires a file path anyway
- It's not clear which of the two (`OpenWithPath` or `FetcherWithPath`) would be responsible for closing the file

Does that sound reasonable?

keegancsmith replied:
We don't need to open the file twice. IE `OpenWithPath` should not open the file, the `FetchWithPath` should do that. If I am not mistaken in the current state for when we invoke fetch:
- Open will always open the file twice
- OpenWithPath will always open the file twice since the fetcher has to write to the path.

I'm suggesting just removing the open from OpenWithPath, since we don't need that first open since the FetchWithPath will _always_ call open.

keegancsmith replied:
ok I totally missed the part where we pass `f` to `copyAndClose`. In that case my comment is incorrect. I think we could move it to  after we have run fetcher ,but this actually feels better to do it before. So nevermind me :) Sorry for the confusion caused.

chrismwendt commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -109,57 +152,214 @@ func filterSymbols(ctx context.Context, symbols []protocol.Symbol, args protocol
+	// `*lowercase_index` enables indexed case insensitive queries.
+	_, err = transaction.Exec(`CREATE INDEX namelowercase_index ON symbols(namelowercase);`)
+	if err != nil {
+		return err
```
I tried it:

- In the transaction but after the insertions
- After the transaction

None of these made a difference in the run time.

chrismwendt commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -73,34 +92,58 @@ func (s *Service) search(ctx context.Context, args protocol.SearchArgs) (result
+	if strings.HasPrefix(regex, "^") && strings.HasSuffix(regex, "$") && regexp.QuoteMeta(unanchoredRegex) == unanchoredRegex {
+		return unanchoredRegex, true
+	}
+	return "", false
```
Any function in particular you'd recommend using from `regexp/syntax`? I don't see a way to check if a string contains any meta characters.

sqs replied:
Check the db package and how it handles RepositoryListOptions’ Query field. There is a func in that pkg that does something similarly.

chrismwendt replied:
I don't see [`RepositoryListOptions` anywhere](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+RepositoryListOptions). This looks close, but not like what you're describing:

https://github.com/sourcegraph/sourcegraph/blob/f3ca04e1ec4a599b780e136085713870af972cfe/cmd/frontend/db/repos.go#L319

Do you have a link?

keegancsmith replied:
I have a lot of experience using syntax.Regexp, so I whipped up this. https://play.golang.org/p/YtTib2ygOzk

chrismwendt replied:
@keegancsmith Perfect! I folded that into this PR and added your tests.

chrismwendt commented on cmd/symbols/internal/symbols/search_test.go:
```diff
@@ -29,6 +32,7 @@ func BenchmarkSearch(b *testing.B) {
 	tests := []protocol.SearchArgs{
 		{Repo: "github.com/sourcegraph/go-langserver", CommitID: "391a062a7d9977510e7e883e412769b07fed8b5e"},
 		{Repo: "github.com/moby/moby", CommitID: "6e5c2d639f67ae70f54d9f2285f3261440b074aa"},
+		{Repo: "github.com/moby/moby", CommitID: "6e5c2d639f67ae70f54d9f2285f3261440b074aa", Query: "^container$"},
```
I split the indexing and searching parts and put the results in the PR description.

chrismwendt commented:
3/3 teammates telling me not to use TypeScript is a pretty strong indicator it's not appropriate, so I rewrote it in bash in https://github.com/sourcegraph/sourcegraph/pull/2334/commits/3ba7dcc852b05a15b9c5f4b810892fbd40aa166d. Here are the main differences:

- The bash version is about half as long
- The TypeScript version gives you a nice stack trace when something fails
- The TypeScript version shows `--help` output
- The bash version had some variables clobbering each other, took ~20 minutes to debug
- It's more clear how values are passed around via functions in TypeScript
- It's easier to navigate the TypeScript version with jump-to-definition
- Prettier is convenient for TypeScript

The lack of type safety and npm ecosystem were less of a pain point than I had expected, but I missed jump-to-definition, objects, and more explicit argument passing.

I imagine that TypeScript would shine if more of our build scripts were written in it, but it'll require one sweeping conversion to avoid the uncanny valley of inconsistency.

chrismwendt commented:
@slimsag Thanks for the review! Unfortunately, `libsqlite3-pcre.{dylib,so}` still must be built, but it happens automatically in `cmd/symbols/build.sh`, and it's cached, so I don't foresee it causing problems.

@keegancsmith Ready for another review :smiley:

felixfbecker commented:
Alright I couldn't resist. Please take a look at this shell script that has objects, exceptions, autocompletion, go-to-definition, autoformatting, local variables, functions with return values and typed explicit parameters, automatic help page, no needed variable quoting, without the syntactic bloat of the TS script: https://gist.github.com/felixfbecker/cce58b198a209abae49780271ae854bd

keegancsmith commented:
> Alright I couldn't resist. Please take a look at this shell script that has objects, exceptions, autocompletion, go-to-definition, autoformatting, local variables, functions with return values and typed explicit parameters, automatic help page, no needed variable quoting, without the syntactic bloat of the TS script: https://gist.github.com/felixfbecker/cce58b198a209abae49780271ae854bd

![image](https://media.giphy.com/media/xTiTnvMb8gkmBvwFiM/giphy.gif)

keegancsmith commented:
> **Benchmark results**

```
$ env CTAGS_COMMAND=$PWD/cmd/symbols/universal-ctags-dev go test -v -bench=. -test.run=sdfsfd ./cmd/symbols/internal/symbols
goos: darwin
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols
BenchmarkSearch/indexing_go-langserver@391-8         	      10	 140234483 ns/op
BenchmarkSearch/indexing_moby@6e5-8                  	       1	7930319865 ns/op
BenchmarkSearch/searching_go-langserver@391_^sortedImportRecord$-8         	   30000	     44872 ns/op
BenchmarkSearch/searching_go-langserver@391_1234doesnotexist1234-8         	    5000	    281762 ns/op
BenchmarkSearch/searching_moby@6e5_^fsCache$-8                             	   20000	     80045 ns/op
BenchmarkSearch/searching_moby@6e5_1234doesnotexist1234-8                  	      30	  46515624 ns/op
PASS
ok  	github.com/sourcegraph/sourcegraph/cmd/symbols/internal/symbols	36.963s
```

> I forced the non exact regexes to be run over all symbols by using `1234doesnotexist1234` as the regex. Exact matches (`^...$`) are much faster and scale better than non exact matches that don't exist.
>
> The current implementation takes ~40ms to search for `1234doesnotexist1234` in moby/moby, and takes ~46ms using sqlite3, which I think is acceptable.

I think it would be more representative to do something like `func.*doesnotex?ist`. PCRE likely optimizes 1234doesnotexist1234 into a substring search which will blaze through. It may get a match, but we still look for a few matches?

Additionally `BenchmarkSearch/indexing_moby@6e5-8` only runs once, (due to it taking 7s). You can do `-benchtime=20s` to force it to run for longer. Given we are interested in memory usage, you can also add `-benchmem`. It would be awesome to copy your tests into master so we can see the difference to the old code (you can compare output with `benchcmp`). But that is just a nice to have. Now that we have the benchmarks we can easily squeeze more juice out by profiling them.

keegancsmith commented on cmd/symbols/internal/symbols/search.go:
```diff
@@ -22,6 +29,18 @@ import (
+func MustRegisterSqlite3WithPcre() {
+	if libSqlite3Pcre == "" {
+		env.PrintHelp()
+		panic("can't find the libsqlite3-pcre library because LIBSQLITE3_PCRE was not set")
```
```suggestion
		log.Fatal("can't find the libsqlite3-pcre library because LIBSQLITE3_PCRE was not set")
```

chrismwendt commented:
@keegancsmith Those would definitely be improvements to make, but they don't **block** merging, do they?

We already know this branch is MUCH better than master. Considering how impactful this change is to all of our browser extension users, and that I've addressed multiple rounds of feedback, I was expecting to get approval by now.

keegancsmith commented on cmd/symbols/internal/symbols/search_test.go:
```diff
@@ -26,19 +35,65 @@ func BenchmarkSearch(b *testing.B) {
+		if err != nil {
+			b.Fatal(err)
+		}
+		defer db.Close()
```
you are still measuring indexing for the search tests. I would suggesting something along the lines of this
```go
search(...) // First search does the indexing
b.ResetTimer() // Only start measuring from this point
for n := 0; n < b.N; n++ {
  search(...)
}
```

chrismwendt replied:
Oh, I thought `b.Run` would start a new timer upon execution of the callback. Fixed.

keegancsmith commented on cmd/symbols/internal/symbols/search_test.go:
```diff
@@ -26,19 +35,65 @@ func BenchmarkSearch(b *testing.B) {
+			b.Fatal(err)
+		}
+		defer db.Close()
+		b.Run(fmt.Sprintf("searching %s@%s %s", path.Base(string(test.Repo)), test.CommitID[:3], test.Query), func(b *testing.B) {
```
These are not nice names. Maybe rather give each test case above a name. Like `regex search moby` or `index moby`.

keegancsmith commented on cmd/symbols/internal/symbols/service_test.go:
```diff
@@ -18,7 +23,58 @@ import (
+		TestCase{Regex: `\$`, WantLiteral: ``, WantOk: false},
+		TestCase{Regex: `\(`, WantLiteral: ``, WantOk: false},
+		TestCase{Regex: `foo$`, WantLiteral: ``, WantOk: false},
+		TestCase{Regex: `(^foo$|^bar$)`, WantLiteral: ``, WantOk: false},
```
you can leave out the wantliteral for the `WantOk: false` cases since the default value is the empty string.

keegancsmith commented on dev/go-install.sh:
```diff
@@ -41,20 +41,20 @@ case $# in
+#             commands="$commands $entCmd "
+#         fi
+#     done
+# fi
```
why is this commented out? Have you tested with ./enterprise/dev/start.sh?

chrismwendt replied:
Oh, I meant to ask @beyang about this. Yes, I've been using ./enterprise/dev/start.sh. @beyang It seems like this code causes the frontend and management-console to be recompiled unnecessarily. I didn't notice any side effects of commenting this out, so I'll remove it unless it serves a purpose.

beyang replied:
@chrismwendt why would this code cause those commands to be rebuilt unnecessarily? If those commands already exist in the list, then this should have no effect. Otherwise, omitting them would mean we are omitting some commands from being built. The former is currently the case, so this shouldn't have any effect on what commands are built/re-built.

chrismwendt replied:
@beyang and I chatted in person, and we concluded that this `if` statement is not necessary and it's correct to remove it (which I did in this PR). The determination of which commands are to be rebuilt upon a file change event happens in `handle-change.sh`.

Asides:

- There are a few duplicated files in OSS and Enterprise builds like `handle-change.sh` and `changewatch.sh` that make it difficult to juggle in your mind
- This seems like the kind of problem that would not exist if we were using a dedicated build system rather than shell scripts. I don't have any suggestions worth mentioning at this point (haven't done much research).

chrismwendt commented on CHANGELOG.md:
```diff
@@ -40,6 +40,9 @@ All notable changes to Sourcegraph are documented in this file.
 - Search queries now use "smart case" by default. Searches are case insensitive unless you use uppercase letters. To explicitly set the case, you can still use the `case` field (e.g. `case:yes`, `case:no`). To explicitly set smart case, use `case:auto`.
+- File match search results now show full repo name if there are results from mirrors on different code hosts (e.g. github.com/sourcegraph/sourcegraph and gitlab.com/sourcegraph/sourcegraph)
+- Search queries now use "smart case" by default. Searches are case insensitive unless you use uppercase letters. To explicitely set the case, you can still use the `case` field (e.g. `case:yes`, `case:no`). To explicitely set smart case, use `case:auto`.
+- Symbols search is much faster now. After the initial indexing, you can expect code intelligence to be nearly instant no matter the size of your repository.
```
Yeah, that's strange, must have been a botched merge. Good catch.

keegancsmith commented on cmd/symbols/internal/symbols/search_test.go:
```diff
@@ -26,13 +33,38 @@ func BenchmarkSearch(b *testing.B) {
+
+	runQueryTest := func(test protocol.SearchArgs) {
+		service.search(ctx, test)
+		b.ResetTimer()
```
sorry I meant you should do the above two lines inside the b.Run. That way if you do `go test -benchrun Search/moby` it will only run the moby tests without doing the extra searches.

chrismwendt commented:
Thanks for all the feedback, @keegancsmith @slimsag @felixfbecker This is a faster, more robust, and more consistent implementation because of you 😄

keegancsmith commented on cmd/symbols/internal/symbols/search_test.go:
```diff
@@ -26,13 +33,41 @@ func BenchmarkSearch(b *testing.B) {
+	runIndexTest := func(test protocol.SearchArgs) {
+		b.Run(fmt.Sprintf("indexing %s@%s", path.Base(string(test.Repo)), test.CommitID[:3]), func(b *testing.B) {
+			for n := 0; n < b.N; n++ {
+				tempFile, err := ioutil.TempFile("", "")
```
missed this in the original review. You need to remove the tempFile once done:
> It is the caller's responsibility to remove the file when no longer needed.

chrismwendt replied:
Ah, good catch, fixed in https://github.com/sourcegraph/sourcegraph/pull/2449
